Added support for Unstoppable Domains

V-Shadbolt opened the pull request:
I've added Unstoppable Domains support alongside the existing ENS implementation. Unstoppable Domains now resolve when entered into the Address search, reverse resolve the connected wallet in the side navbar, and reverse resolve within the message previews.

The Bearer token for the API calls was omitted from the PR (listed as JWT) for security purposes. I wasn't successfully able to pass env variables through to the call so would like to work with the team on your API key practices and triage the PR accordingly.

```javascript
const response = await fetch(
      `https://resolve.unstoppabledomains.com/reverse/${address.toLowerCase()}`,
      {
        headers: {
          Authorization: "Bearer JWT",
        },
      },
    );
```

![messages](https://github.com/xmtp-labs/xmtp-inbox-web/assets/28935974/50429e4b-ac65-48c2-9117-36acc45f7bc9)
![navbar](https://github.com/xmtp-labs/xmtp-inbox-web/assets/28935974/14e8d628-dd1b-4a78-8517-3b407eb912f4)

daria-github commented on component-library/components/InfoCard/InfoCard.stories.tsx:
```diff
@@ -33,7 +33,7 @@ InfoCardNoArrow.args = {
   header: "Send a new message",
   subtext:
-    "Find an existing contact or message someone using their wallet address or ENS address",
+    "Find an existing contact or message someone using their wallet address or domain address",
```
thanks for updating the copy in the stories!

daria-github commented on helpers/string.ts:
```diff
@@ -35,11 +35,56 @@ export const isEnsAddress = (address: string): boolean => {
+  return ALLOWED_UNS_SUFFIXES.some((suffix) => address.endsWith(suffix));
+};
+
+export const fetchUnsName = async (address: any): Promise<string> => {
```
we could add the key as an env var in Vercel

V-Shadbolt replied:
Done! Shoutout to Yash in TG for providing me the Vercel var name.

daria-github commented on helpers/tests/string.test.ts:
```diff
@@ -61,6 +62,25 @@ describe("isEnsAddress", () => {
   });
 });
 
+describe("isUnsAddress", () => {
```
great tests! nit for the 2nd test: "should return false if address does not end with any UNS suffix"

V-Shadbolt replied:
Nice catch! Agreed and implemented.

daria-github commented on locales/en_US.json:
```diff
@@ -28,14 +28,14 @@
     "messages_empty_subheader": "Get started",
     "messages_empty_cta_1_header": "Send a message",
-    "messages_empty_cta_1_subheader": "Message someone using their 0x wallet or ENS address",
+    "messages_empty_cta_1_subheader": "Message someone using their 0x wallet or domain address",
```
tagging @jhaaaa here to weigh in on the suggested copy change

V-Shadbolt replied:
I wasn't sure how you wanted to handle this as i know there's a lot more languages than just en. Happy to roll this back until we find a more sutable solution but I would like to include UNS in the placeholder text in some capacity to draw attention to the new functionality. My thoughts were 
```
Enter a 0x wallet or domain address
``` 
or 
```
Enter a 0x wallet, ENS, or UNS address
```

With the former being more agnostic and scalable.

jhaaaa replied:
Thank you for the tag, @daria-github and @V-Shadbolt. <3

I love the idea of explicitly surfacing UNS in the placeholder text, as suggested:

`Enter a 0x wallet, ENS, or UNS address`

I think at this point in the lifecycle, it is helpful to be explicit about what's supported for user clarity and education. To me, we can do this until the UI gets too cluttered and makes things more difficult to understand. But hopefully by that time, we won't have to be explicit because all major address formats will be supported and usage patterns will be more mainstream.

daria-github commented on wrappers/AddressInputWrapper.tsx:
```diff
@@ -47,6 +49,22 @@ export const AddressInputWrapper = () => {
+  useEffect(() => {
+    const getUns = async () => {
+      if (isValidLongWalletAddress(recipientWalletAddress)) {
+        const name = await fetchUnsName(recipientWalletAddress);
```
what're your thoughts on chaining a catch here & adding some error handling?

V-Shadbolt replied:
Did you have some thoughts on error handling here? I've adjusted the logic on the API calls themselves to match the ENS logic within the WAGMI connector but if you have some other thoughts more than happy to implement! We could look at a try:catch where we default to setting name to null

I know fetch doesn't catch HTTP 401 and 403 errors for example so I can always move to something like axios to improve these.

daria-github replied:
ah, makes sense to align that. i think we can hold off adding anything here for the time being.

daria-github commented on wrappers/AddressInputWrapper.tsx:
```diff
@@ -56,11 +74,13 @@ export const AddressInputWrapper = () => {
-          (size[0] < 700
-            ? shortAddress(recipientWalletAddress)
-            : recipientWalletAddress),
+        displayAddress: ensName
```
since this nested ternary is getting a bit long, WDYT of moving this to its own var?

V-Shadbolt replied:
I've fixed this up in the files were any overly-nested ternary as I completely agree - it wasn't super legible. I've thrown in a new const that handles the preliminary check. 
```javascript
const domain = ensName ?? unsName;
```

daria-github replied:
great idea; much cleaner. thanks!

daria-github reviewed (commented):
Thanks for submitting this pr @V-Shadbolt & fantastic job on it! 👏 Just left a few small comments; once we add the JWT we can deploy this to Vercel and do some testing of this feature.

V-Shadbolt commented:
Thanks @daria-github ! I've implemented all the fixes minus the additional error handling on the useEffect calls as I'm curious as to your thoughts. Will never say no to additional error handling! 

I also managed to catch a bug that I had missed on the initial submission. The below line was missing the UNS reference. When i fixed the ternary, I included the new const `domain`in place of `ensName`
```javascript
walletAddress: ensName ? recipientWalletAddress : "",
```

I also changed the typings / returns of the UD API calls to match the ENS logic within the WAGMI SDK.

daria-github commented:
Thanks for the fast turnaround with addressing the comments @V-Shadbolt! Code looks great & tests/checks are passing; I've gone ahead and deployed this so we can test the change. We can explore those possible copy changes or the additional error handling during testing. Will ping you here if we run into any changes requested. Thanks again for your work on this!

daria-github commented on wrappers/MessagePreviewCardWrapper.tsx:
```diff
@@ -38,6 +39,18 @@ export const MessagePreviewCardWrapper = ({
+  const [previewUnsName, setPreviewUnsName] = useState<string | null>();
+
+  useEffect(() => {
+    const getUns = async () => {
```
could we make this call only if the address is of UNS format? (same in SideNav)

V-Shadbolt replied:
Definitely. An easy check that should cut down excess network calls. Great catch! I'm AFK at the moment but I'll add that logic in alongside any other comments you have!

V-Shadbolt replied:
Hey @daria-github ! For the preview and side nav, I added `isValidLongWalletAddress()` but I don't know of an explicit way to check if the wallet address is UNS aside from resolving it.

daria-github replied:
Makes sense — if there's no pattern, we can leave as-is. Thanks for checking!

yash-luna commented:
Latest preview link does not hav reverse resolution. Conversation started with alohajha.x displays the 0x address after refreshing the app

![image](https://github.com/xmtp-labs/xmtp-inbox-web/assets/118013479/5f790d7a-9049-4f3a-a9b0-082552fd7f6a)

daria-github commented:
@V-Shadbolt just a quick FYI — noticed an issue where the first message for the UNS addresses I tried didn't appear to go through until a refresh; this had to do with the conversation IDs of these addresses being lowercase and the mapped conversations having the address checksum'd. Made a fix for this on the dev branch and merged this branch with dev. PR for reference: https://github.com/xmtp-labs/xmtp-inbox-web/pull/274

@yash-luna are you still seeing the reverse resolution issue? I'm seeing this work properly for the same address on the preview link.

V-Shadbolt commented:
Hey @daria-github , @yash-luna ! I've added some logic per the above comment and can't replicate the reverse resolution issue. 

I've also gone ahead and removed UNS from `FullMessageWrapper.tsx`. Doing so reduced the network calls when changing conversations from 5 to 1. I couldn't see any UI impacts from this change and it seems to be localized to state. Happy to put it back in but it's a huge reduction in calls.

daria-github reviewed (approved):
@V-Shadbolt I just made the copy changes to this branch which was the only pending change, rest looks great - going to approve and merge. Thanks again for your work on this!
